llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -3049,9 +3242,29 @@ void AddressSanitizer::initializeCallbacks(Module &M, const TargetLibraryInfo *T
                                     GlobalVariable::NotThreadLocal, 1);
         });
 
-    AsanSetShadowDeviceLocalFunc =
-        M.getOrInsertFunction("__asan_set_shadow_local_memory", IRB.getVoidTy(),
+    // __asan_set_shadow_static_local(
+    //   uptr ptr,
+    //   size_t size,
+    //   size_t size_with_redzone
+    // )
+    AsanSetShadowStaticLocalFunc =
+        M.getOrInsertFunction("__asan_set_shadow_static_local", IRB.getVoidTy(),
                               IntptrTy, IntptrTy, IntptrTy);
+
+    // __asan_set_shadow_dynamic_local(
+    //   uptr ptr,
+    //   uint32_t num_args
+    // )
+    AsanSetShadowDynamicLocalFunc = M.getOrInsertFunction(
+        "__asan_set_shadow_dynamic_local", IRB.getVoidTy(), IntptrTy, Int32Ty);
+
+    AsanLaunchInfo = M.getOrInsertGlobal(
+        "__AsanLaunchInfo", IntptrTy->getPointerTo(kSpirOffloadGlobalAS), [&] {
+          return new GlobalVariable(
+              M, IntptrTy->getPointerTo(kSpirOffloadGlobalAS), false,
+              GlobalVariable::ExternalLinkage, nullptr, "__AsanLaunchInfo",
```

Comment:
Do we run this pass on every translation unit, or do we run it after we have linked everything together? If the former, then this variable will create multiple definitions error at link stage

Reply:
> Do we run this pass on every translation unit, or do we run it after we have linked everything together? 

We run this pass on every TU.

> If the former, then this variable will create multiple definitions error at link stage

This variable is external symbol, it only defined in libdevice.